compiler-hacking.md

```diff
@@ -0,0 +1,502 @@
+
+All this implicit capturing can produce some pretty _yikes_ methods, but if you
+look at lambda-heavy Java code, the methods also tend to be pretty yikes.
+
+There's a constant trade off here: more complexity in the compiler can be used
+to avoid capturing unnecessary values.
+
+The `Group` and `LeftJoin` operations are particularly gnarly since they have
+multiple expressions connected together. For instance, consider:
+
+    Group
+      ys = Where z List y + a
+    By x
+
+If `x`, `y` and `z` are all stream variables, then only `a` is captured.
+
+For grouping, two functions are needed: one which makes a key to split the data
+up into subsets, and then one that merges everything in a subset together.
+
+For ease, the compiler will generate a single set of captures for both of those
+operations. That means that the key operation will capture `a` even though it
+doesn't need it.
+
+## Compiler Mechanics
+
+The compiler operates in a few phases:
+
+1. Parse the file into a parse tree, rooted as a `ProgramNode`.
+2. Resolve input formats, actions, functions, and `Define`/call references.
+3. Resolve variables.
+4. Check the type and stream purity of everything.
+5. Convert everything to bytecode. This is in the _render_ methods.
+
+There are a lot of classes in the compiler, but the major ones are:
+
+- [`OliveNode`](shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/OliveNode.java), which declares each top-level element in the file (function, olive, define olive, constant)
+- [`OliveClauseNode`](shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/OliveClauseNode.java), which defines clauses in an olive
+- [`ExpressionNode`](shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/compiler/ExpressionNode.java), which defines all the bits and pieces of expressions
+
+For details on components, the JavaDoc outlines the behaviour and expectations
+of the classes. Most of the documentation is on the abstract classes and the
+concrete class have less since they just execute the behaviour defined by the
+abstract class.
+
+For general naming conventions:
+- _resolve_ methods convert strings into useful metadata (_e.g._ find the [`FunctionDefinition`](shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/FunctionDefinition.java) given the name in the script)
+- _render_ methods generate bytecode
+- classes that end in _Node_ are generate type of syntax in the language; their subclasses represent concrete bits of syntax the user can write
+- classes that end in _Builder_ are utility classes that make generating bytecode simpler
+
+`ExpressionNode` has many similar clones. For instance, there is an expression
+node for `For` and then ones for each of the sources, clauses, subsamples, and
+collectors in a `For` expression, containing other expressions.  While these
+are structurally different, they are built very similarly to `ExpressionNode`.
+
+### Type Information
+Computing captures is done as part of bytecode production. The type system for
```

Comment:
`Computing captures is done as part of bytecode production` <- feels kinda disjointed from both the section header and the next sentences, could it be given more context or moved?